### PR TITLE
Geocoded Firms and Subsidiaries

### PR DESCRIPTION
**Pull request by lshepstone**
This drives the work being done in the `geocoding` branch for `rad`.

### CODE REVIEW

File: app/jobs/geocode_firm_job.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class GeocodeFirmJob < ActiveJob::Base
+  def perform(firm)
+    results = Geocoder.search(firm.full_street_address)
+
+    if results.any?
+      Stats.increment('radsignup.geocode.firm.success')
```
Comment (benlovell) [inline app/jobs/geocode_firm_job.rb]: Maybe we can push this into a helper here and use `stat :success` or `stat :failure` and push the duplication of keys down?

File: app/jobs/geocode_firm_job.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+      firm.latitude = results.first.latitude
```
Comment (benlovell) [inline app/jobs/geocode_firm_job.rb]: As discussed in person - this is a great candidate for pushing into the domain. `Firm#geocode!` or some such. I suggested a bang method as it mutates your `Firm` instance. But I'm easy on that.

File: app/jobs/geocode_firm_job.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      firm.longitude = results.first.longitude
+      firm.save(callbacks: false)
+    else
+      Stats.increment('radsignup.geocode.firm.failed')
+
+      firm.latitude = nil
```
Comment (benlovell) [inline app/jobs/geocode_firm_job.rb]: Obviously the same applies here as the prior comment.

File: spec/models/firm_spec.rb [hunk 4]
```diff
@@ -185,4 +187,67 @@
...
+    context 'when the address is present' do
+      it 'the firm is scheduled for geocoding' do
+        expect(GeocodeFirmJob).to receive(:perform_later).with(firm)
+        firm.save!
+      end
+    end
+
+    context 'when the address has changed' do
+      let(:firm) do
+        create(:firm).tap do |f|
```
Comment (benlovell) [inline spec/models/firm_spec.rb]: This can just be `create(:firm, address_postcode: 'ABCD 123')`

File: app/jobs/geocode_firm_job.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class GeocodeFirmJob < ActiveJob::Base
+  def perform(firm)
+    results = Geocoder.search(firm.full_street_address)
+
+    if results.any?
+      Stats.increment('radsignup.geocode.firm.success')
+      firm.geocode!(results.first.latitude, results.first.longitude)
+    else
+      Stats.increment('radsignup.geocode.firm.failed')
+      firm.geocode!(nil, nil)
```
Comment (benlovell) [inline app/jobs/geocode_firm_job.rb]: How about `Firm#geocode(lat = nil, long = nil)` so this case golfs down to `firm.geocode!`

File: app/models/firm.rb [hunk 6]
```diff
@@ -86,6 +86,36 @@ class Firm < ActiveRecord::Base
...
+    write_attribute(:latitude, value)
+  end
+
+  def longitude=(value)
+    value = value.to_f.round(6) unless value.nil?
+    write_attribute(:longitude, value)
+  end
+
+  def geocode!(latitude, longitude)
+    begin
```
Comment (benlovell) [inline app/models/firm.rb]: I'd favour just saving and skipping callbacks if that's the intention.

File: app/models/firm.rb [hunk 7]
```diff
@@ -118,6 +148,16 @@ def field_order
 
   private
 
+  def geocode_in_progress?
```
Comment (benlovell) [inline app/models/firm.rb]: Obviously we could drop this if you go with the prior comment's recommendations.

**Comment by benlovell** [comment]
:+1:

**Comment by alexwllms** [comment]
:+1:

**Pull request merged**